ql/src/java/org/apache/hadoop/hive/ql/parse/type/ExprFactory.java

```diff
@@ -0,0 +1,361 @@
+   */
+  protected abstract T createDoubleConstantExpr(String value);
+
+  /**
+   * Creates a decimal constant expression from input value.
+   * If the constant created from the input value is null, we return:
+   * 1) a constant expression containing null value if allowNullValueConstantExpr is true, or
+   * 2) null if allowNullValueConstantExpr is false.
+   */
+  protected abstract T createDecimalConstantExpr(String value, boolean allowNullValueConstantExpr);
+
+  /**
+   * Creates a string constant expression from input value.
+   */
+  protected abstract T createStringConstantExpr(String value);
+
+  /**
+   * Creates a date constant expression from input value.
+   */
+  protected abstract T createDateConstantExpr(String value);
+
+  /**
+   * Creates a timestamp constant expression from input value.
+   */
+  protected abstract T createTimestampConstantExpr(String value);
+
+  /**
+   * Creates a timestamp with local time zone constant expression from input value.
+   * ZoneId is the local time zone.
+   */
+  protected abstract T createTimestampLocalTimeZoneConstantExpr(String value, ZoneId zoneId);
+
+  /**
+   * Creates a interval year-month constant expression from input value.
+   */
+  protected abstract T createIntervalYearMonthConstantExpr(String value);
+
+  /**
+   * Creates a interval day-time constant expression from input value.
+   */
+  protected abstract T createIntervalDayTimeConstantExpr(String value);
+
+  /**
+   * Creates a interval year constant expression from input value.
+   */
+  protected abstract T createIntervalYearConstantExpr(String value);
+
+  /**
+   * Creates a interval month constant expression from input value.
+   */
+  protected abstract T createIntervalMonthConstantExpr(String value);
+
+  /**
+   * Creates a interval day constant expression from input value.
+   */
+  protected abstract T createIntervalDayConstantExpr(String value);
+
+  /**
+   * Creates a interval hour constant expression from input value.
+   */
+  protected abstract T createIntervalHourConstantExpr(String value);
+
+  /**
+   * Creates a interval minute constant expression from input value.
+   */
+  protected abstract T createIntervalMinuteConstantExpr(String value);
+
+  /**
+   * Creates a interval second constant expression from input value.
+   */
+  protected abstract T createIntervalSecondConstantExpr(String value);
+
+  /**
+   * Default generator for constant expression when type cannot be inferred
+   * from input query.
+   */
+  protected T createConstantExpr(String value) {
+    // The expression can be any one of Double, Long and Integer. We
+    // try to parse the expression in that order to ensure that the
+    // most specific type is used for conversion.
+    T result = null;
+    T result2 = null;
+    try {
+      result = createDoubleConstantExpr(value);
+      if (value != null && !value.toLowerCase().contains("e")) {
+        result2 = createDecimalConstantExpr(value, false);
```

Comment:
If createDecimalConstantExpr throws a NumberFormatException then the function won't try to parse the value as Bigint and Int, so I think it should be surrounded with try/catch.

Reply:
What you describe above is the correct behavior: this code is already in Hive in `TypeCheckProcFactory`.

If an exception is thrown for decimal, we consider the constant to be a double. If an exception is not thrown for decimal, we try to parse as bigint and int. If any of those succeed, we will use those. But if none of them do, we will use the decimal.

If we want to refactor this crazy logic, I would rather create a new JIRA for that instead of doing it in this JIRA.